### PR TITLE
Simplify go.mod

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Stops adding explicit `replace` directives for all dependencies. This was needed in the early days of go.mod to ensure we had a consistent root view of the required dependency versions and to keep spurious changes from happening to go.mod, but both of those issues have been resolved since go1.17.

A nice side effect is that dependency updates should no longer require several pin/update-vendor/lint cycles.

This also fixes up the unwanted-dependencies scanner to work with the expanded `go mod graph` output that results from dropping the forced pins to specific versions.

Requires depstat fix in https://github.com/kubernetes-sigs/depstat/pull/65 before CI will pass on this PR

```release-note
NONE
```

/cc @dims
/area code-organization

### CODE REVIEW

File: hack/unwanted-dependencies.json [hunk 1]
```diff
@@ -23,9 +23,17 @@
     }
   },
   "status": {
```
Comment (liggitt) [inline hack/unwanted-dependencies.json:25]: this now shows not just which unwanted dependencies are still in use, but what they are referenced by, so we can see if that is growing or what upstreams we need to work with to eliminate calls

Comment (dims) [reply hack/unwanted-dependencies.json:25]: Nice!

File: go.mod [hunk 2]
```diff
@@ -247,278 +247,6 @@ require (
...
-	gopkg.in/gcfg.v1 => gopkg.in/gcfg.v1 v1.2.0
-	gopkg.in/inf.v0 => gopkg.in/inf.v0 v0.9.1
-	gopkg.in/natefinch/lumberjack.v2 => gopkg.in/natefinch/lumberjack.v2 v2.0.0
-	gopkg.in/square/go-jose.v2 => gopkg.in/square/go-jose.v2 v2.2.2
-	gopkg.in/warnings.v0 => gopkg.in/warnings.v0 v0.1.1
-	gopkg.in/yaml.v2 => gopkg.in/yaml.v2 v2.4.0
-	gopkg.in/yaml.v3 => gopkg.in/yaml.v3 v3.0.1
-	gotest.tools/v3 => gotest.tools/v3 v3.0.3
-	honnef.co/go/tools => honnef.co/go/tools v0.0.1-2020.1.4
 	k8s.io/api => ./staging/src/k8s.io/api
```
Comment (dims) [inline go.mod:250]: cool, so we are left with just staging directories.

**Comment by dims** [comment]
I like it! good direction @liggitt. thanks!

**Comment by liggitt** [comment]
rebased, fixed lint error with unused method

it looks like check-dependency-stats has the same issue with the expanded graph, will dig into that

**Comment by liggitt** [comment]
depstat fix in https://github.com/kubernetes-sigs/depstat/pull/65

**Pull request merged**